src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs

```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+            }
+
+            return false;
         }
 
-        public Settings(string root, string fileName, bool isMachineWideSettings)
+        public bool SetItemInSection(string sectionName, SettingsItem item, bool isBatchOperation = false)
         {
-            if (string.IsNullOrEmpty(root))
+            if (string.IsNullOrEmpty(sectionName))
             {
-                throw new ArgumentException("root cannot be null or empty", nameof(root));
+                throw new ArgumentNullException(nameof(sectionName));
             }
 
-            if (string.IsNullOrEmpty(fileName))
+            if (item == null)
             {
-                throw new ArgumentException(Resources.Argument_Cannot_Be_Null_Or_Empty, nameof(fileName));
+                throw new ArgumentNullException(nameof(item));
             }
 
-            if (!FileSystemUtility.IsPathAFile(fileName))
+            // Check if set is an update
+            if (_computedSections.TryGetValue(sectionName, out var section))
             {
-                throw new ArgumentException(Resources.Settings_FileName_Cannot_Be_A_Path, nameof(fileName));
+                if (section.TryUpdateChildItem(item, isBatchOperation))
+                {
+                    return true;
+                }
             }
 
-            Root = root;
-            FileName = fileName;
-            XDocument config = null;
-            ExecuteSynchronized(() => config = XmlUtility.GetOrCreateDocument(CreateDefaultConfig(), ConfigFilePath));
-            ConfigXDocument = config;
-            IsMachineWideSettings = isMachineWideSettings;
-            CheckConfigRoot();
+            var outputSettingsFile = GetOutputSettingsFileForSection(sectionName);
+
+            // Set is an add
+            return outputSettingsFile != null && outputSettingsFile.RootElement.AddItemInSection(sectionName, item.Copy(), isBatchOperation);
         }
 
         public event EventHandler SettingsChanged = delegate { };
 
-        public IEnumerable<ISettings> Priority
+        public Settings(string root)
+            : this(new SettingsFile(root)) { }
+
+        public Settings(string root, string fileName)
+            : this(new SettingsFile(root, fileName)) { }
+
+        public Settings(string root, string fileName, bool isMachineWide)
+            : this(new SettingsFile(root, fileName, isMachineWide)) { }
+
+        private Settings(SettingsFile settingsHead)
+        {
+            _settingsHead = settingsHead;
+
+            var curr = _settingsHead;
+            while (curr != null)
+            {
+                curr.SettingsChanged += (_, __) => { RefreshSettingValues(); };
+                curr = curr.Next;
+            }
+
+            MergeSettingFilesValues();
+        }
+
+        private void RefreshSettingValues()
+        {
+            MergeSettingFilesValues();
+            SettingsChanged?.Invoke(this, EventArgs.Empty);
```

Comment:
`SettingsChanged` event should not be raised here since you haven't saved the file yet.

Reply:
This method is only called from the `SettingsChanged` of a specific `SettingsFile`, I could remove this method and only make it inline in the event handler so that no one can call this without changing a file.